test/integration/replicaset/replicaset_test.go

```diff
@@ -272,246 +280,595 @@ func TestAdoption(t *testing.T) {
-	waitRSStable(t, clientSet, rs, ns.Name)
+	// test to ensure extensions/v1beta1 selector is mutable
+	newSelectorLabels := map[string]string{"changed_name_extensions_v1beta1": "changed_test_extensions_v1beta1"}
+	rs.Spec.Selector.MatchLabels = newSelectorLabels
+	rs.Spec.Template.Labels = newSelectorLabels
+	replicaset, err := clientSet.ExtensionsV1beta1().ReplicaSets(ns.Name).Update(rs)
+	if err != nil {
+		t.Fatalf("failed to update extensions/v1beta1 replicaset %s: %v", replicaset.Name, err)
+	}
+	if !reflect.DeepEqual(replicaset.Spec.Selector.MatchLabels, newSelectorLabels) {
+		t.Errorf("selector should be changed for extensions/v1beta1, expected: %v, got: %v", newSelectorLabels, replicaset.Spec.Selector.MatchLabels)
+	}
 
-	// change the rs's selector to match both pods
-	patch := `{"spec":{"selector":{"matchLabels": {"uniqueKey":null}}}}`
-	rsClient := clientSet.Extensions().ReplicaSets(ns.Name)
-	rs, err := rsClient.Patch(rs.Name, types.StrategicMergePatchType, []byte(patch))
+	// test to ensure apps/v1beta2 selector is immutable
+	rsV1beta2, err := clientSet.AppsV1beta2().ReplicaSets(ns.Name).Get(replicaset.Name, metav1.GetOptions{})
 	if err != nil {
-		t.Fatalf("Failed to patch replica set: %v", err)
+		t.Fatalf("failed to get apps/v1beta2 replicaset %s: %v", replicaset.Name, err)
 	}
-	t.Logf("patched rs = %#v", rs)
-	// wait for the rs select both pods and delete one of them
-	if err := wait.Poll(10*time.Second, 60*time.Second, func() (bool, error) {
-		return verifyRemainingObjects(t, clientSet, ns.Name, 1, 1)
+	newSelectorLabels = map[string]string{"changed_name_apps_v1beta2": "changed_test_apps_v1beta2"}
+	rsV1beta2.Spec.Selector.MatchLabels = newSelectorLabels
+	rsV1beta2.Spec.Template.Labels = newSelectorLabels
+	_, err = clientSet.AppsV1beta2().ReplicaSets(ns.Name).Update(rsV1beta2)
+	if err == nil {
+		t.Fatalf("failed to provide validation error when changing immutable selector when updating apps/v1beta2 replicaset %s", rsV1beta2.Name)
+	}
+	expectedErrType := "Invalid value"
+	expectedErrDetail := "field is immutable"
+	if !strings.Contains(err.Error(), expectedErrType) || !strings.Contains(err.Error(), expectedErrDetail) {
+		t.Errorf("error message does not match, expected type: %s, expected detail: %s, got: %s", expectedErrType, expectedErrDetail, err.Error())
+	}
+}
+
+// Update .Spec.Replicas to x and verify .Status.Replicas is changed accordingly
+func testStatusReplicas(t *testing.T, c clientset.Interface, rs *v1beta1.ReplicaSet, ns string, x int32) {
+	rsClient := c.Extensions().ReplicaSets(ns)
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		newRS, err := rsClient.Get(rs.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		*(rs.Spec.Replicas) = x
+		_, err = rsClient.Update(newRS)
+		return err
 	}); err != nil {
-		t.Fatal(err)
+		t.Fatalf("Failed to update .Spec.Replicas to %d for rs %s: %v", x, rs.Name, err)
 	}
-	close(stopCh)
+	waitRSStable(t, c, rs, ns)
```

Comment:
We have to wait after the update to ensure `.Status.Replicas` is equal to `.Spec.Replicas`.